src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Doctrine;
+
+use Doctrine\DBAL\Connection as DriverConnection;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\DBAL\Schema\Synchronizer\SingleDatabaseSynchronizer;
+use Doctrine\DBAL\Types\Type;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class Connection
+{
+    private $tableName;
+    private $driverConnection;
+    private $debug;
+
+    const STATUS_PENDING = 'pending';
+    const STATUS_PROCESSING = 'processing';
+    const STATUS_ACK = 'ack';
+
+    public function __construct(string $tableName, DriverConnection $driverConnection, bool $debug = false)
+    {
+        $this->tableName = $tableName;
+        $this->driverConnection = $driverConnection;
+        $this->debug = $debug;
+    }
+
+    public function publish(string $body, array $headers): void
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+        $this->driverConnection->insert($this->tableName, array(
+            'body' => $body,
+            'headers' => \json_encode($headers),
+        ));
+    }
+
+    public function get(): ?array
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $doctrineEnvelope = $this->driverConnection->createQueryBuilder()
+            ->select('m.*')
+            ->from($this->tableName, 'm')
+            ->where('m.status = :status')
+            ->orderBy('id', 'ASC')
+            ->setMaxResults(1)
+            ->setParameter(':status', self::STATUS_PENDING)
+            ->execute()
+            ->fetch();
+
+        if (false === $doctrineEnvelope) {
+            return null;
+        }
+
+        $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+        $this->changeStatus($doctrineEnvelope['id'], self::STATUS_PROCESSING);
+
+        return $doctrineEnvelope;
+    }
+
+    public function ack($id): bool
+    {
+        return $this->changeStatus($id, self::STATUS_ACK);
+    }
+
+    public function nack($id): bool
+    {
+        return $this->changeStatus($id, self::STATUS_PENDING);
```

Comment:
The message may be stuck in status `STATUS PROCESSING` if the connection was closed, for example due to restart server/DB

Reply:
Maybe we should delete the message after retrieve it. And reinsert on nack. Then the `ack` method can be removed.

Reply:
@vtsykun How can we avoid that ? I'm thinking of adding a command to change status of messages in `processing` state for long time. So adding a `updated_at` field.

Do you have another idea ?